### PR TITLE
Ingester Main

### PR DESCRIPTION
**Pull request by davit-y**
## Which problem is this PR solving?
- #929 

## Short description of the changes
- Add options.go to configure ingester using viper
- Add main.go to initialize and start Ingester

### CODE REVIEW

File: cmd/ingester/app/options.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+	"github.com/spf13/viper"
+
+	"github.com/jaegertracing/jaeger/pkg/kafka/consumer"
+	"github.com/jaegertracing/jaeger/plugin/storage/kafka"
+)
+
+const (
+	configPrefix      = "ingester"
+	suffixBrokers     = ".brokers"
```
Comment (black-adder) [inline cmd/ingester/app/options.go]: hmmm, I'm not sure about this. These configs shouldn't belong on the ingester itself, but should be encapsulated inside the storage factory. Do we have to do this like this since we haven't come up with a better story for the storage factory and flags?

Comment (davit-y) [reply cmd/ingester/app/options.go]: We talked in person, but I'll recap for those who may be curious. Early in the design process there was a discussion about wether the Ingester should operate on an abstract queue or for Kafka. We decided it will be designed for kafka, but be independent of the write storage. Therefore there isn't an issue with having kafka specific config on the Ingester as long as the write storage is handled similarly to the other jaeger applications.

File: cmd/ingester/main.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+			storageFactory.InitFromViper(v)
+			if err := storageFactory.Initialize(baseFactory, logger); err != nil {
+				logger.Fatal("Failed to init storage factory", zap.Error(err))
+			}
+			spanWriter, err := storageFactory.CreateSpanWriter()
+			if err != nil {
+				logger.Fatal("Failed to create span writer", zap.Error(err))
+			}
+			unmarshaller := app.UnmarshallerFromType(options.Encoding)
```
Comment (black-adder) [inline cmd/ingester/main.go]: I can see us copy pasting a lot of this kafka initialization logic when we OSS the indexer (is that part of the plan btw). Do you think it'll make sense to create something like https://github.com/jaegertracing/jaeger/blob/master/pkg/cassandra/config/config.go to handle all the kafka bootstrapping rather than dumping it all here in main?

Comment (black-adder) [reply cmd/ingester/main.go]: hmm, let me rephrase a bit, internally, we don't want to write this same main logic if we can help it. Ideally, we create a builder for the ingester similar to the agent builder https://github.com/jaegertracing/jaeger/blob/master/cmd/agent/app/builder.go and internally, we can just run builder.build() instead of copy pasting this same code.

File: cmd/ingester/app/options.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	suffixTopic       = ".topic"
+	suffixGroupID     = ".group-id"
+	suffixParallelism = ".parallelism"
+	suffixEncoding    = ".encoding"
+
+	encodingJSON  = "json"
+	encodingProto = "protobuf"
+
+	defaultBroker      = "127.0.0.1:9092"
+	defaultTopic       = "jaeger-ingester-spans"
```
Comment (vprithvi) [inline cmd/ingester/app/options.go]: Is this the default topic used by the `jaeger-collector`'s kafka writers to emit spans?

File: cmd/ingester/main.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+				logger.Fatal("Unable to set up consumer", zap.Error(err))
+			}
+			kafkaConsumer.Start()
+
+			hc.Ready()
+			select {
+			case <-signalsChannel:
+				err := kafkaConsumer.Close()
+				if err != nil {
+					logger.Error("Failed to close span writer", zap.Error(err))
```
Comment (vprithvi) [inline cmd/ingester/main.go]: Does closing the consumer also close the span writer?

File: cmd/ingester/main.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+				if err != nil {
+					logger.Error("Failed to close consumer", zap.Error(err))
+				}
+				if closer, ok := spanWriter.(io.Closer); ok {
+					err := closer.Close()
+					if err != nil {
+						logger.Error("Failed to close span writer", zap.Error(err))
+					}
+				}
+				logger.Info("Jaeger Ingester is finishing")
```
Comment (vprithvi) [inline cmd/ingester/main.go]: nit: I'd also add a log line on L101 stating "Beginning shutdown of Jaeger Ingester" or something similar.

**Comment by davit-y** [comment]
Redoing using a builder to allow for easy initialization of an ingester

File: cmd/ingester/main.go (hunk 2 above)
Comment (davit-y) [reply cmd/ingester/main.go]: Done

File: cmd/ingester/app/builder/builder.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	// DefaultBroker is the default kafka broker
+	DefaultBroker = "127.0.0.1:9092"
+	// DefaultTopic is the default kafka topic
+	DefaultTopic = "jaeger-spans"
+	// DefaultGroupID is the default consumer Group ID
+	DefaultGroupID = "jaeger-ingester"
+	// DefaultParallelism is the default parallelism for the span processor
+	DefaultParallelism = 1000
+	// DefaultEncoding is the default span encoding
+	DefaultEncoding = EncodingJSON
```
Comment (vprithvi) [inline cmd/ingester/app/builder/builder.go]: I'm curious, is this what we emit from the collectors? (Not protobuf?)

File: cmd/ingester/app/builder/builder.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	Parallelism int
+	Encoding    string
+}
+
+// CreateConsumer creates a new span consumer for the ingester
+func (b *Builder) CreateConsumer(logger *zap.Logger, metricsFactory metrics.Factory, spanWriter spanstore.Writer) (*consumer.Consumer, error) {
+	var unmarshaller kafka.Unmarshaller
+	if b.Encoding == EncodingJSON {
+		unmarshaller = kafka.NewJSONUnmarshaller()
+	} else {
```
Comment (vprithvi) [inline cmd/ingester/app/builder/builder.go]: I'm not a fan of this approach because if someone passes in an arbitrary encoding, they would get a protobuf marshaller. I recommend checking for the type and returning an error if nothing matches.

File: cmd/ingester/app/builder/builder.go (hunk 6 above)
Comment (davit-y) [reply cmd/ingester/app/builder/builder.go]: My mistake, default should be protobuf
